src/compiler/file_uploader/on_before_exit.ts

```diff
@@ -1,41 +1,81 @@
+import { createClearFileAndInfoActor } from './clear_file_and_info_actor';
 import { getListOfIncompleteFiles } from './incomplete_files';
+import {
+    FileInfo,
+    getOngoingHashes,
+    reportOnGoingHashes
+} from './ongoing_hashes';
 
 export function onBeforeExit(canisterId: string, paths: [Src, Dest][]) {
     let complete = false;
-    let endAttempts = 0;
+    let cleanUpComplete = false;
+    let ongoingFileHashes: FileInfo[] = [];
     process.on('beforeExit', async (code) => {
-        endAttempts++;
-        if (complete) {
+        if (cleanUpComplete) {
             // If any async behavior happens in 'beforeExit' then 'beforeExit'
             // will run again. This is need to prevent an infinite loop
             return;
         }
-        complete = await verifyUploadAndHashingComplete(
+        if (complete) {
+            await cleanup(canisterId, paths);
+            cleanUpComplete = true;
+            return;
+        }
+        ongoingFileHashes = await verifyUploadAndHashingComplete(
             canisterId,
             paths,
-            false // TODO end after a number of attempts?
+            ongoingFileHashes
         );
+
+        complete = ongoingFileHashes.length === 0;
+
+        console.info(`Waiting 5 seconds and then we'll try again.`);
         await new Promise((resolve) => setTimeout(resolve, 5000));
     });
 }
 
+async function cleanup(canisterId: string, paths: [Src, Dest][]) {
+    const incompleteFiles = await getListOfIncompleteFiles(paths, canisterId);
+    for (const [_, path] of incompleteFiles) {
+        const actor = await createClearFileAndInfoActor(canisterId);
+        await actor.clear_file_and_info(path);
+    }
+}
+
 async function verifyUploadAndHashingComplete(
     canisterId: string,
     paths: [Src, Dest][],
-    finished: boolean
-): Promise<boolean> {
-    let incompleteFiles = await getListOfIncompleteFiles(paths, canisterId);
-    if (incompleteFiles.length === 0 || finished) {
-        return true;
+    previousHashInfos: FileInfo[]
+): Promise<FileInfo[]> {
+    const incompleteFiles = await getListOfIncompleteFiles(paths, canisterId);
+    if (incompleteFiles.length === 0) {
+        return [];
     } else {
-        console.log(
-            `Missing hashes for ${
-                incompleteFiles.length
-            } files:\n${incompleteFiles.join(
-                '\n'
-            )}. Waiting 5 seconds and then we'll try again.`
+        const incompleteDestPaths = incompleteFiles.map(([_, path]) => path);
+
+        const ongoingHashInfo = await getOngoingHashes(
+            canisterId,
+            previousHashInfos,
+            incompleteDestPaths
+        );
+
+        const wasUpdated = ongoingHashInfo.every(
+            (fileInfo) => fileInfo.triesSinceLastChange < 5
         );
-        return false;
+
+        if (!wasUpdated) {
+            // TODO get hashing percentage and report, if not update after 5 times end the process
```

Comment:
Let's do this

Reply:
It was already done, I just need to remove the TODO